File: frontend/src/app/api/registration.ts
import api from 'codeforlife/lib/esm/api';

const registrationApi = api.injectEndpoints({
  endpoints: (build) => ({
    requestIndependentStudentPasswordReset: build.mutation<
      null,
      {
        email: string;
      }
    >({
      query: (body) => ({
        url: 'user/password/reset/student/',
        method: 'POST',
        body,
        headers: {
          'Content-Type': 'application/x-www-form-urlencoded'
        }
      })
    }),
    requestTeacherPasswordReset: build.mutation<
      null,
      {
        email: string;
      }
    >({
      query: (body) => ({
        url: 'user/password/reset/teacher/',
        method: 'POST',
        body,
        headers: {
          'Content-Type': 'application/x-www-form-urlencoded'
        }
      })
    }),
    resetPassword: build.mutation<
      null,
      {
        userId: string;
        token: string;
        body: {
          new_password1: string;
          new_password2: string;
        };
      }
    >({
      query: ({ userId, token, body }) => ({
        url: `user/password/reset/${userId}-${token}/`,
        method: 'POST',
        body,
        headers: {
          'Content-Type': 'application/x-www-form-urlencoded'
        }
      })
    }),
    verifyPassword: build.mutation<
      { isPasswordCorrect: string },
      {
        password: string;
      }
    >({
      query: (body) => ({
        url: 'verify-password/',
        method: 'POST',
        body,
        headers: {
          'Content-Type': 'application/x-www-form-urlencoded'
        }
      })
    }),
    deleteAccount: build.mutation<
      null | {
        password?: string;
      },
      {
        password: string;
        unsubscribeNewsletter: boolean;
      }
    >({
      query: (body) => ({
        url: 'delete/account/',
        method: 'POST',
        body,
        headers: {
          'Content-Type': 'application/x-www-form-urlencoded'
        }
      })
    })
  })
});

export default registrationApi;
export const {
  useRequestIndependentStudentPasswordResetMutation,
  useRequestTeacherPasswordResetMutation,
  useResetPasswordMutation,
  useVerifyPasswordMutation,
  useDeleteAccountMutation
} = registrationApi;

File: backend/api/urls/registration.py
from django.urls import path, re_path

from ..views.registration import (
    student_password_reset,
    teacher_password_reset,
    password_reset_check_and_confirm,
    delete_account,
    verify_password,
)


urlpatterns = [
    path(
        "user/password/reset/student/",
        student_password_reset,
        name="student_password_reset",
    ),
    path(
        "user/password/reset/teacher/",
        teacher_password_reset,
        name="teacher_password_reset",
    ),
    re_path(
        r"^user/password/reset/(?P<uidb64>[0-9A-Za-z]+)-(?P<token>.+)/$",
        password_reset_check_and_confirm,
        name="password_reset_check_and_confirm",
    ),
    path(
        "delete/account/",
        delete_account,
        name="delete_account",
    ),
    path("verify-password/", verify_password, name="verify_password"),
]
